### PR TITLE
Add object det and referring expression det on ScanNet

### PR DESCRIPTION
**Pull request by alexsax**
## Motivation and Context

Add representation evaluation on object detection and referring expression on ScanNet bounding boxes.

## Changelog

- Created .py script that demos creating a SparseVoxelMap on ScanNet data, and evaluates object detection.
- Added option to do mask erosion to instance masks before unprojecting in SparseVoxelMap
- Fixed some shape bugs in creating VoxelizedPointcloud when no RGB image present

## How Has This Been Tested

Visualized results on a sample scannet scene using different SparseVoxelGrid hyperparameters and visually checked correspondence with quantitative results. Hyperparameters varied: iou_association_threshold, 

## Types of changes
- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [x] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+            boxes_plot_together=False,
+            boxes_name_int_to_display_name_dict=dict(
+                enumerate(self.metadata.thing_classes)
+            ),
+            backend="pytorch3d",
+        )
+
+
+# TODO: use this code to create semantic sensor
+# def create_semantic_sensor(
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:114]: Dead code or add an issue?

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+#     vocab = build_vocab_from_category_map(obj_name_to_id, rec_name_to_id)
+#     semantic_sensor.update_vocabulary_list(vocab, 0)
+#     semantic_sensor.set_vocabulary(0)
+#     return config, semantic_sensor
+
+
+if __name__ == "__main__":
+    from eval.utils import eval_bboxes_and_print
+
+    parser = argparse.ArgumentParser(
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:137]: I've been using click, i dont know if we care to make it consistent? this is fine

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        help="Display the scene in an Open3D visualizer window",
+    )
+    args = parser.parse_args()
+    torch_device = "cpu" if args.device == -1 else f"cuda:{args.device}"
+
+    # Load specific scene
+    data = ScanNetDataset(
+        root_dir=args.scannet_dir,
+        frame_skip=args.frame_skip,
+        # referit3d_config = ReferIt3dDataConfig(),
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:186]: Dead code or add a flag maybe?

Comment (cpaxton) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:186]: maybe htis is in config somewhere?

File: projects/scannet_offline_eval/eval/refer_exp.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+# Return:
+# - IoU and IoU @ thresh
+# - Against target
+# - Against target class (is it getting distracted by other objects of the same class)
+# - Against all objects (is it localizing any object?)
+
+
+@torch.no_grad()
+def get_stats_ious_at_thresh(
+    ious: Tensor,
```
Comment (cpaxton) [inline projects/scannet_offline_eval/eval/refer_exp.py:38]: We can probably move a lot of these to home_robot/datasets/eval?

File: src/home_robot/home_robot/datasets/scannet/scannet_dataset.py [hunk 5]
```diff
@@ -27,8 +27,8 @@ class ScanNetDataset(object):
     # Segmentation data
     METAINFO = {
         "classes": (
-            "wall",
-            "floor",
+            # "wall",
```
Comment (cpaxton) [inline src/home_robot/home_robot/datasets/scannet/scannet_dataset.py:30]: why?

File: src/home_robot/home_robot/utils/image.py [hunk 6]
```diff
@@ -2,15 +2,22 @@
...
+import torch.nn.functional as F
 import trimesh.transformations as tra
+from torch import Tensor
 
 
 class Camera(object):
-    """Simple pinhole camera model. Contains parameters for projecting from depth to xyz, and saving information about camera position for planning."""
+    """
+    Simple pinhole camera model. Contains parameters for projecting from depth to xyz, and saving information about camera position for planning.
+    TODO: Move this to utils/cameras.py?
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/image.py:19]: yeah probably a good idea

Comment (cpaxton) [reply src/home_robot/home_robot/utils/image.py:19]: utils could use some cleanup tbh

File: src/home_robot/home_robot/utils/image.py [hunk 7]
```diff
@@ -193,3 +200,53 @@ def rotate_image(imgs: List[np.ndarray]) -> List[np.ndarray]:
...
+        masks = [(t >= val - tol) & (t <= val + tol) for t in masks]
+        mask = functools.reduce(lambda a, b: a & b, masks).unsqueeze(1)
+    else:
+        mask = (target >= val - tol) & (target <= val + tol)
+    mask = 0 != F.conv2d(
+        mask.float(),
+        torch.ones(1, 1, mask_extra_radius, mask_extra_radius, device=mask.device),
+        padding=(mask_extra_radius // 2),
+    )
+    #     mask = F.conv2d(mask.float(), torch.ones(1, 1, 5, 5, device=mask.device), padding=2) != 0
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/image.py:232]: dead code

File: src/home_robot/home_robot/utils/image.py [hunk 8]
```diff
@@ -193,3 +200,53 @@ def rotate_image(imgs: List[np.ndarray]) -> List[np.ndarray]:
...
+    target: Tensor, val: float = 0.0, tol: float = 1e-3, mask_extra_radius: int = 5
+) -> Tensor:
+    """Build mask where all channels are (val - tol) <= target <= (val + tol)
+        Optionally, dilate by mask_extra_radius
+
+    Args:
+        target (Tensor): [B, N_channels, H, W] input tensor
+        val (float, optional): Value to use for masking. Defaults to 0.0.
+        tol (float, optional): Tolerance for mask. Defaults to 1e-3.
+        mask_extra_radius (int, optional): Dilate by mask_extra_radius pix . Defaults to 5.
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/image.py:215]: if its optional should we check for none

File: src/home_robot/home_robot/utils/image.py [hunk 9]
```diff
@@ -193,3 +200,53 @@ def rotate_image(imgs: List[np.ndarray]) -> List[np.ndarray]:
...
+
+    Returns:
+        _type_: Mask of shape target.shape
+    """
+    assert target.ndim == 4, target.shape
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/image.py:220]: better error message here, maybe f"target should be of shape [B, N_channels, H, W], was {target.shape}"

File: src/home_robot/home_robot/utils/image.py [hunk 10]
```diff
@@ -193,3 +200,53 @@ def rotate_image(imgs: List[np.ndarray]) -> List[np.ndarray]:
...
+    mask = 0 != F.conv2d(
+        mask.float(),
+        torch.ones(1, 1, mask_extra_radius, mask_extra_radius, device=mask.device),
+        padding=(mask_extra_radius // 2),
+    )
+    #     mask = F.conv2d(mask.float(), torch.ones(1, 1, 5, 5, device=mask.device), padding=2) != 0
+    return (~mask).expand_as(target)
+
+
+def dilate_or_erode_mask(mask: Tensor, radius: int, num_iterations=1) -> Tensor:
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/image.py:236]: description please

File: src/home_robot/home_robot/utils/image.py [hunk 11]
```diff
@@ -193,3 +200,53 @@ def rotate_image(imgs: List[np.ndarray]) -> List[np.ndarray]:
...
+    erode = radius < 0
+    if erode:
+        mask = ~mask
+    mask = mask.half()
+    conv_kernel = torch.ones(
+        (1, 1, abs_radius, abs_radius), dtype=mask.dtype, device=mask.device
+    )
+    for _ in range(num_iterations):
+        mask = mask.half()
+        mask = F.conv2d(mask, conv_kernel, padding="same")
```
Comment (cpaxton) [inline src/home_robot/home_robot/utils/image.py:248]: i like that we're using torch here instead of cv

**Review by cpaxton — COMMENTED** [review]
some minor comments but essentially lgtm

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 12]
```diff
@@ -0,0 +1,355 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+    Placeholder code to create 3D map from a scannet scene
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:6]: Should I not review this then?

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+from home_robot.utils.voxel import VoxelizedPointcloud
+
+
+class SemanticVocab(IntEnum):
+    FULL = auto()
+    SIMPLE = auto()
+    ALL = auto()
+
+
+class SparseVoxelMapAgent:
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:46]: This agent should probably move into home_robot/mapping or home_robot/datasets/scannet right?

File: src/home_robot/home_robot/mapping/voxel.py [hunk 14]
```diff
@@ -106,14 +118,16 @@ def reset(self) -> None:
         self.observations = []
         # Create an instance memory to associate bounding boxes in space
         self.instances = InstanceMemory(
-            num_envs=1, du_scale=1, instance_association="bbox_iou"
+            num_envs=1,
+            **self.instance_memory_kwargs,
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/voxel.py:122]: Should we have these default args set somewhere?

File: src/home_robot/home_robot/mapping/voxel.py [hunk 15]
```diff
@@ -234,7 +248,7 @@ def add(
 
         # Add instance views to memory
         instance = instance_image.clone()
-        instance[~valid_depth] = -1
+        # instance[~valid_depth] = -1
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/voxel.py:251]: dead code/how should we handle this? why is this removed?

File: src/home_robot/home_robot/mapping/voxel.py [hunk 16]
```diff
@@ -137,7 +151,7 @@ def reset_cache(self):
 
     def get_instances(self) -> List[InstanceView]:
         """Return a list of all viewable instances"""
-        return tuple(self.instances.instance_views[0].values())
+        return tuple(self.instances.instances[0].values())
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/voxel.py:154]: Is this still returning InstanceView?

**Review by cpaxton — COMMENTED** [review]
minor issues

File: src/home_robot/home_robot/mapping/voxel.py (hunk 14 above)
Comment (alexsax) [reply src/home_robot/home_robot/mapping/voxel.py:122]: Good idea (though num_envs will always be 1 here)

File: src/home_robot/home_robot/mapping/voxel.py (hunk 15 above)
Comment (alexsax) [reply src/home_robot/home_robot/mapping/voxel.py:251]: If we apply valid_depth at this stage, then instance_image is 1d (n_kept) instead of 2d (h, w)
We probably want to keep the instance mask and the valid mask around, since we want the instance mask for masking out the RGB and both of them for masking out the pointcloud

Comment (alexsax) [reply src/home_robot/home_robot/mapping/voxel.py:251]: This was removed because we handle the masking later, in process_instances_for_env

File: src/home_robot/home_robot/utils/image.py (hunk 6 above)
Comment (alexsax) [reply src/home_robot/home_robot/utils/image.py:19]: Yes but in another commit :)

File: src/home_robot/home_robot/utils/image.py (hunk 8 above)
Comment (alexsax) [reply src/home_robot/home_robot/utils/image.py:218]: You're right -- not optional in the sense that it can't be none.

File: src/home_robot/home_robot/utils/image.py (hunk 11 above)
Comment (alexsax) [reply src/home_robot/home_robot/utils/image.py:251]: We should be using it everywhere :)

File: projects/scannet_offline_eval/eval/refer_exp.py (hunk 4 above)
Comment (alexsax) [reply projects/scannet_offline_eval/eval/refer_exp.py]: I'm thinking:
- Would it make sense to move it within the home_robot/datasets/scannet/eval?
- Do we plan to evaluate other datasets in home_robot
-  Which would be easier to port out of home_robot?

File: projects/scannet_offline_eval/build_sparse_voxel_map.py (hunk 1 above)
Comment (alexsax) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py]: Dead code :)

File: projects/scannet_offline_eval/build_sparse_voxel_map.py (hunk 2 above)
Comment (alexsax) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:318]: Yeah I think we should use click or hydra for this. Hydra probably makes it easier to maintain/test

File: projects/scannet_offline_eval/build_sparse_voxel_map.py (hunk 3 above)
Comment (alexsax) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:367]: Moved to hydra config -- will push changes soon

File: src/home_robot/home_robot/mapping/voxel.py (hunk 15 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:251]: I think it's still useful to handle it earlier here, for navigation?

Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:251]: I guess this actually is irrelevant to that though so sgtm

File: src/home_robot/home_robot/mapping/voxel.py [hunk 17]
```diff
@@ -442,31 +457,42 @@ def show(
...
-            # features = [colors],
-            names=[torch.stack(names, dim=0).unsqueeze(-1)],
-        )
+        if instances:
+            bounds, names = zip(
+                *[(v.bounds, v.category_id) for v in self.get_instances()]
+            )
+            detected_boxes = BBoxes3D(
+                bounds=[torch.stack(bounds, dim=0)],
+                # features = [colors],
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/voxel.py:488]: dead code?

Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:488]: should we be passing color?

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 18]
```diff
@@ -0,0 +1,475 @@
...
+            instance_scores=instance_scores,
+        )
+
+        if visualize_map:
+            # Now draw 2d
+            self.voxel_map.get_2d_map(debug=True)
+
+    def step_trajectory(
+        self, obs_list: Sequence[Observations], cache_key: Optional[str] = None
+    ):
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:137]: This could use a description

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 19]
```diff
@@ -0,0 +1,475 @@
...
+            # load from cache
+            assert self.cache_dir is not None
+            raise NotImplementedError
+        else:
+            for i, obs in enumerate(obs_list):
+                self.step(obs)
+            if self.cache_dir is not None:
+                # Save to cache
+                raise NotImplementedError
+        print(f"Found {len(self.voxel_map.get_instances())} instances")
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:148]: should it print this every time?

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 20]
```diff
@@ -0,0 +1,475 @@
...
+                instance=None,  # These could be cached
+                # Pose of the camera in world coordinates
+                camera_pose=scene_obs["poses"][i],
+                camera_K=scene_obs["intrinsics"][i],
+                task_observations={
+                    # "features": scene_obs["images"][i],
+                },
+            )
+            obs_list.append(obs)
+        self.step_trajectory(obs_list)
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:203]: I added support for this in demo branch

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 21]
```diff
@@ -0,0 +1,475 @@
...
+
+        # Do NMS on global bboxes
+        if self.global_nms_thresh > 0.0 and len(self.voxel_map.get_instances()) > 0:
+            # Loop this since when we combine bboxes, they are now bigger.
+            # And some bounding boxes might be inside the new larger bboxes
+            last_n_instances = -1
+            while len(self.voxel_map.get_instances()) != last_n_instances:
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:210]: Would this be useful to break out into a helper function in utils? Maybe that's too much

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 22]
```diff
@@ -0,0 +1,475 @@
...
+            height=1000,
+            boxes_plot_together=False,
+            boxes_name_int_to_display_name_dict=dict(
+                enumerate(self.metadata.thing_classes)
+            ),
+            backend="pytorch3d",
+        )
+
+
+def compress_instance_bounds_(
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:260]: move utils?

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 23]
```diff
@@ -0,0 +1,475 @@
...
+            voxel_size=voxel_size,
+            min_points_after_drop=min_voxels_to_compress,
+        )
+        new_bounds = get_bounds(reduced_points)
+        if box3d_volume_from_bounds(new_bounds) >= min_vol:
+            instance.bounds = new_bounds
+    return instances
+
+
+def drop_smallest_weight_points(
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:281]: move utils + docstring?

File: projects/scannet_offline_eval/build_sparse_voxel_map.py (hunk 2 above)
Comment (cpaxton) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:318]: hydra is probably better i agree

Comment (cpaxton) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:318]: we don't need to do this now though

File: projects/scannet_offline_eval/build_sparse_voxel_map.py [hunk 24]
```diff
@@ -0,0 +1,475 @@
...
+    from home_robot.perception.detection.detic.detic_perception import DeticPerception
+
+    parser = argparse.ArgumentParser(
+        prog="build_3d_map",
+        description="Builds 3D map of a scannet scene from a RGBD trajectory and evaluates object detection",
+    )
+
+    parser.add_argument(
+        "--scannet-dir",
+        default="/private/home/ssax/home-robot/src/home_robot/home_robot/datasets/scannet/data",
```
Comment (cpaxton) [inline projects/scannet_offline_eval/build_sparse_voxel_map.py:325]: hard coded path? can we do a relative path instead? usually i have these coded to assume youre running from $HOME_ROBOT_ROOT

File: src/home_robot/home_robot/datasets/spot/data_utils.py [hunk 25]
```diff
@@ -0,0 +1,342 @@
...
+import argparse
+import dataclasses
+import datetime
+import pickle as pkl
+import sys
+import timeit
+from pathlib import Path
+from typing import Optional, Sequence, Tuple, Union
+
+import bosdyn.client.frame_helpers as frame_helpers
```
Comment (cpaxton) [inline src/home_robot/home_robot/datasets/spot/data_utils.py:16]: we probably want this in home_robot_spot not in home_robot proper?
I guess it will only be lazily imported for now, but can you file an issue for when we merge spot code?

File: src/home_robot/home_robot/datasets/spot/data_utils.py [hunk 26]
```diff
@@ -0,0 +1,342 @@
...
+        compass=None,
+        rgb=color,
+        depth=depth,
+        camera_pose=cam_to_world,
+        camera_K=camera.K[:3, :3],
+        task_observations={"timestamp": timestamp},
+    )
+
+
+class SpotPickleDataset:
```
Comment (cpaxton) [inline src/home_robot/home_robot/datasets/spot/data_utils.py:161]: What script creates this? Maybe a light description? not totally sure, its fairly obvious

File: src/home_robot/home_robot/datasets/spot/data_utils.py [hunk 27]
```diff
@@ -0,0 +1,342 @@
...
+# This can all go in __main__
+###############################
+if __name__ == "__main__":
+    depth_key = "frontleft_depth_in_visual_frame"
+    rgb_key = "frontleft_fisheye_image"
+    rgb_video_file_name = "traj.mp4"
+    use_sparse_voxel_map_agent = False
+    OUTPUT_PLY_FILE = "traj.ply"
+    DATA_DIR = Path(
+        "/private/home/ssax/home-robot/src/home_robot/home_robot/datasets/spot/data"
```
Comment (cpaxton) [inline src/home_robot/home_robot/datasets/spot/data_utils.py:209]: relative path?

**Review by cpaxton — COMMENTED** [review]
only minor comments. We should pull explict xyz_frame code from demo branch

File: src/home_robot/home_robot/mapping/voxel.py (hunk 17 above)
Comment (alexsax) [reply src/home_robot/home_robot/mapping/voxel.py:488]: At some point we can color the boxes according to class, but that's not implemented yet :)

Comment (alexsax) [reply src/home_robot/home_robot/mapping/voxel.py:488]: This is sort of a placeholder for that

Comment (alexsax) [reply src/home_robot/home_robot/mapping/voxel.py:488]: I'll just change the comment

File: src/home_robot/home_robot/mapping/voxel.py (hunk 16 above)
Comment (alexsax) [reply src/home_robot/home_robot/mapping/voxel.py:154]: Good catch!

File: projects/scannet_offline_eval/build_sparse_voxel_map.py (hunk 21 above)
Comment (alexsax) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:210]: Good idea -- I moved this to InstanceMemory to be clearer

File: projects/scannet_offline_eval/build_sparse_voxel_map.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (cpaxton) [reply projects/scannet_offline_eval/build_sparse_voxel_map.py:182]: @alexsax add xyz_frame?

File: src/home_robot/home_robot/mapping/voxel.py [hunk 28]
```diff
@@ -479,31 +664,76 @@ def _show_pytorch3d(self, **plot_scene_kwargs):
...
-                },
-                # Could add keyframes or instances here.
-            },
+            plots={"Global scene": traces},
             **update(_default_plot_args, plot_scene_kwargs),
         )
         return fig
 
-    def _show_open3d(self, instances: bool, **backend_kwargs):
+    def sample_explored(self) -> Optional[np.ndarray]:
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/voxel.py:672]: Did these move here from the SVMNavigationSpace?

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py [hunk 29]
```diff
@@ -61,28 +66,74 @@ class InstanceView:
...
     """
     A single instance found in the environment. Each instance is composed of a list of InstanceView objects, each of which is a view of the instance at a particular timestep.
     """
 
-    def __init__(self):
-        """
-        Initialize Instance
+    name: str = None
+    category_id: int = None
+    """ingeger indicating the category"""
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:77]: @alexsax spelling thing

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py [hunk 30]
```diff
@@ -263,21 +422,139 @@ def find_global_instance_by_bbox_overlap(
...
+            local_instance_view_id (int): The local instance view ID whose global counterpart needs to be found.
+
+        Returns:
+            Optional[int]: The global instance ID with the most point cloud overlap.
+                        Returns None if no such global instance is found.
+
+        TODO:
+            - Optimize by having global instances store a voxelized point cloud to keep the number of points manageable.
+        """
+        raise NotImplementedError
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:529]: whats going on here?

**Review by cpaxton — APPROVED** [review]
LGTM, just some minor comments. can file issues for anything left unresolved
